### DISCUSSION TITLE
Failed to Evaluate Dynamic Condition in Rule Using DynamicExpresso

### DISCUSSION THREAD

**Discussion by venktesh786** [NRules/NRules#364] [Q&A]
**Issue:** I am using NRules to evaluate dynamic conditions in my rule engine. The rules are defined dynamically, and I am utilizing DynamicExpresso to evaluate the conditions at runtime. However, I am encountering the following exception during rule execution, specifically when calling session.Insert(entity):

**Exception:**

> RuleLhsExpressionEvaluationException: Failed to evaluate condition
holdable => value(CustomerRule.DynamicHoldableRule).EvaluateCondition(holdable, value(CustomerRule.DynamicHoldableRule).RuleData.Condition)

In my base class for dynamic rule generation, the model and conditions are dynamic. I have created an interface, IHoldable, with the necessary properties, which can be inherited by all models.

`public interface IHoldable
{
    bool OnHold { get; set; }
    bool IsChecked { get; set; }
    string Name { get; }
} 

**Dynamic Rule Generation Base Class`**

`public class DynamicHoldableRule : Rule
{
    public RuleData RuleData { get; set; } // This contains the dynamic condition and action

    public override void Define()
    {
        IHoldable holdable = null;

        When()
            .Match<IHoldable>(() => holdable, p => EvaluateCondition(p, RuleData.Condition));

        Then()
            .Do(ctx => ExecuteAction(holdable, RuleData.Action));
    }

    private bool EvaluateCondition(IHoldable holdable, string condition)
    {
        if (holdable == null || string.IsNullOrEmpty(condition))
        {
            Console.WriteLine("Invalid holdable or condition.");
            return false;
        }

        try
        {
            var interpreter = new Interpreter();
            return interpreter.SetVariable("p", holdable).Eval<bool>(condition);
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error evaluating condition: {ex.Message}");
            return false;
        }
    }

    private void ExecuteAction(IHoldable holdable, string action)
    {
        if (string.IsNullOrEmpty(action)) return;

        if (action == "HoldOrder")
        {
            Console.WriteLine($"{holdable.Name}'s orders are on hold.");
        }
        else
        {
            Console.WriteLine($"{holdable.Name}'s orders are not on hold.");
        }
    }
}

**RuleData Class`**

`public class RuleData
{
    public string Condition { get; set; } 
    public string Action { get; set; }
    public string RuleName { get; set; } 
}
` 

**Customer Class**

`public class Customer : IHoldable
{
    public int CustomerId { get; set; }
    public string CustomerName { get; set; }
    public bool OnHoldCustomer { get; set; }
    public string Name => CustomerName;
}`

**Program Class**

'class Program
{
    static void Main(string[] args)
    {
        // Create some customer objects
        var customers = new List<Customer>
        {
            new Customer { CustomerId = 1, CustomerName = "Customer A", OnHoldCustomer = true },
            new Customer { CustomerId = 2, CustomerName = "Customer B", OnHoldCustomer = false }
        };

        // Define dynamic rule data
        var ruleData = new List<RuleData>
        {
            new RuleData { Condition = "OnHoldCustomer == true", Action = "HoldOrder", RuleName = "On Hold Rule" }
        };

        var ruleFactory = new RuleFactory();

        // Create custom rule loader
        var customRuleLoader = new CustomRuleLoader(ruleFactory);

        // Create RuleEngine and execute rules
        var ruleEngine = new RuleEngine(new RuleRepository().Compile(), customRuleLoader);
        ruleEngine.RunRules(customers.Cast<IHoldable>().ToList(), ruleData);
    }
}
'

**Steps Taken:**

1. **Dynamic Condition:**
       The condition "p.OnHoldCustomer == true" is stored in RuleData.Condition and is evaluated at runtime using DynamicExpresso.

2. **Entity Insertion:**
   The error occurs during the session.Insert(entity) call when inserting customer objects.

3. **Dynamic Rule Evaluation:**
    The EvaluateCondition method uses DynamicExpresso to dynamically evaluate the condition by passing the holdable (customer) 
    object.

**Question:** How can I resolve the "Failed to evaluate condition" error in NRules when using dynamic condition evaluation with DynamicExpresso? Is there a better approach to debug this issue or handle dynamic conditions in NRules more effectively?

For your reference, I have attached my POC Console project.
[CustomerRule.zip](https://github.com/user-attachments/files/17072293/CustomerRule.zip)

**Comment by venktesh786** [comment]
I have referred CustomRuleRepository in sample project in the repo and fixed it

**Comment by snikolayev** [comment]
@venktesh786 the problem is that in your `CustomRuleLoader` you are instantiating a rule, but then instead of using that instance, you are loading the rule from the type, using `RuleRepository.Load`.
If instead of logging exception.Message, you log the whole exception, you'll see the stack trace and a NullReferenceException, since the rule loaded from the type does not have RuleData initialized.

Use this in your `CustomRuleLoader`:

```c#
        public void LoadRules(RuleRepository repository, List<RuleData> rules)
        {
            var ruleSet = new RuleSet("CustomRuleSet");
            var ruleDefinitionFactory = new RuleDefinitionFactory();

            foreach (var ruleData in rules)
            {
                // Create a dynamic rule instance with injected RuleData
                var ruleInstance = _ruleFactory.CreateRule(ruleData);
                var ruleDefinition = ruleDefinitionFactory.Create(ruleInstance);
                ruleSet.Add(ruleDefinition);
            }
            repository.Add(ruleSet);
        }
```

**Comment by snikolayev** [comment]
On a more general note, if you are trying to build your own expression evaluator on top of NRules, I would not use the Fluent DSL layer, and just directly generate rules in the canonical rules form (see https://nrules.net/articles/advanced/rule-builder.html#:~:text=Rule%20builder%20is%20a%20component,RuleModel%20namespace.)
The way you have it, DSL just adds an unnecessary layer of complexity.
